internal/utils/archive_extract.go

```diff
@@ -0,0 +1,103 @@
+package utils
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/ossf/package-analysis/internal/log"
+)
+
+// ExtractTarGzFile extracts a .tar.gz / .tgz file located at path, optionally
+// accompanied by a temporary change to the given directory during extraction.
+func ExtractTarGzFile(path string, chdir string) error {
+	return processGzipFile(path, func(reader io.Reader) error {
+		return extractTar(reader, chdir)
+	})
+}
+
+func processGzipFile(path string, process func(io.Reader) error) error {
+	gzFile, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	var unzippedBytes *gzip.Reader
+	if unzippedBytes, err = gzip.NewReader(gzFile); err != nil {
+		return err
+	}
+
+	defer func() {
+		if err = unzippedBytes.Close(); err != nil {
+			log.Error("failed to close gzip reader: %w", err)
+		}
+	}()
+
+	if err = process(unzippedBytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// extractTar extracts the contents of the given stream of bytes of a tar archive.
+// If chdir is non-empty, the working directory will be temporarily changed to the
+// specified one before the extraction takes place, so that relative paths are resolved
+// relative to that directory. Otherwise, relative paths will be resolved relative to
+// the current working directory.
+func extractTar(tarStream io.Reader, chdir string) error {
+	var header *tar.Header
+	var err error
+
+	if chdir != "" {
```

Comment:
Rather than changing dir and modifying global state, can we just support an `outputDir` directory that is written to without a chdir call? I.e. through computing the paths  through path.Join as we unpack.

Also, let's have callers be explicit and not support an empty `outputDir`.

Reply:
Yeah I agree changing global state isn't ideal, I just took inspiration from `tar -C`. I was also thinking that if the archive has absolute paths then these should be preserved, but actually on second thought they probably shouldn't in our case. So yes, given all this, let's do what you said.